### DISCUSSION TITLE
Двойное условие

### DISCUSSION THREAD

**Discussion by vldekov** [totumonline/totum-ru-issues-and-discussions#301] [Как это сделать?] [answered]
Не могу понять почему не работает следующий код. Помогите разобраться. =:setList(table: 't_glavnaya'; field: 'status' = "Не действует КР"; where: cond`'nasrokdo' < $#nd && 'status' != "КР верифицирована"`; log: true)

**Comment by vldekov** [comment, marked as answer]
=:setList(table: 't_glavnaya'; field: 'status' = "Не действует КР"; where: 'nasrokdo' < $#nd; where: 'status' != "КР верифицирована"; log: true)
вот так работает )

**Reply by ttm-alexeypolunin** [reply]
Да, `cond` в `where` не пашет. Мы обсудили и пока остановились на варианте:

```
=: setList(table: 't_glavnaya'; field: 'status' = "Не действует КР"; where: qrow`'nasrokdo' < $#nd && 'status' != "КР верифицирована"`; log: true)
``` 

Напишу, когда будет работать.

**[Работает с 4.8.52.1](https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/301#discussioncomment-4786862)**

**Reply by vldekov** [reply]
Супер! Теперь с новым параметром wcond возможны варианты ИЛИ ( || ) ?

**Reply by ttm-alexeypolunin** [reply]
Да, он будет решать эту задачу, когда мы это сделаем.

**Comment by ttm-alexeypolunin** [comment]
### QROW в where-параметрах

> С версии 4.8.52.1

В where параметрах функций может быть использована специальная конструкция `qrow`, которая позволяет задать различные условия поиска для БД.

```totum
=: select(table: $#ntn; field: 'id'; where: qrow`'field_1' > 100 || 'field_2' = true`)
```

`qrow` поддерживает операторы:

+ `&&` — И

+ `||` — ИЛИ

+ `()` — группировка

Операторы сравнения аналогичны операторам остальной системы.

Name-параметры указываются в одинарных кавычках `'field_1'`

В сравнении `'field_1' > 100` обязательно должен быть использован один name-параметр.

Name-параметры могут быть использованы в обоих сторонах сравнения `'field_1' > 'field_2'` — выбрать строки где `field_1` > `field_2`.

Пример с группировкой:

```totum
=: select(table: $#ntn; field: 'id'; where: qrow`'(field_1' > 100 && 'field_3' = true) || 'field_2' = true`)

// field_1 больше 100 и field_3 = true или field_2 равно true

```

На одном уровне **операторы должны быть одинаковыми!**

Это ошибочный пример:

```totum
=: select(table: $#ntn; field: 'id'; where: qrow`'(field_1' > 100 && 'field_3' = true) || 'field_2' = true && filed_4 = false`)

// Результат ОШБК! тк на верхнем уровне используются II и &&

```

Правильно будет:

```totum
=: select(table: $#ntn; field: 'id'; where: qrow`'((field_1' > 100 && 'field_3' = true) || 'field_2' = true) && filed_4 = false`)

// (field_1 больше 100 и field_3 = true или field_2 равно true) и filed_4 = false

```

По прежнему может быть использовано несколько `where`, они работают в режиме `И`:


```totum
=: select(table: $#ntn; field: 'id'; where: qrow`'(field_1' > 100 && 'field_3' = true) || 'field_2' = true`; where: 'filed_4' = false)

// (field_1 больше 100 и field_3 = true или field_2 равно true) и filed_4 = false

```

> ВАЖНО!

Если у вас в сравнениях используется `id` в режиме `И` то запрос будет исполняться **сильно быстрее**, если поставить это условие в отдельный where вперед

```totum
=: select(table: $#ntn; field: 'id'; where: 'id' != #id; where: qrow`'(field_1' > 100 && 'field_3' = true) || 'field_2' = true`)

// id не равно id текущей строки и (field_1 больше 100 и field_3 = true или field_2 равно true)

```
